### PR TITLE
Side exercises for tournament part1

### PR DESCRIPTION
**Pull request by glennj**
| exercise | canonical data | difficulty |
| --- | --- | :-: |
| [book-store](https://github.com/glennj/tcl/blob/side-exercises-for-tournament-part1/exercises/book-store/README.md) | [canonical-data.json](https://github.com/exercism/problem-specifications/blob/master/exercises/book-store/canonical-data.json) | 8 |
| [connect](https://github.com/glennj/tcl/blob/side-exercises-for-tournament-part1/exercises/connect/README.md) | [canonical-data.json](https://github.com/exercism/problem-specifications/blob/master/exercises/connect/canonical-data.json) | 7 |
| [react](https://github.com/glennj/tcl/blob/side-exercises-for-tournament-part1/exercises/react/README.md) | [canonical-data.json](https://github.com/exercism/problem-specifications/blob/master/exercises/react/canonical-data.json) | 8 |

Part 1 of 2.

These are 3 of the hardest exercises.

book-store was translated from [Java](https://github.com/exercism/java/blob/master/exercises/book-store/.meta/src/reference/java/BookStore.java).

### CODE REVIEW

File: exercises/book-store/example.tcl [hunk 1]
```diff
@@ -0,0 +1,94 @@
+namespace eval BookStore {
+    namespace export cost
+
+    # cost in cents
+    variable bookPrice 800
```
Comment (kotp) [inline exercises/book-store/example.tcl:5]:
```suggestion
    variable bookPrice 800
```

File: exercises/book-store/example.tcl [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+            set newGroupBooks [lrange $distinctBooks 0 $i]
+            set costOfNewGroup [costOfGroupSize [llength $newGroupBooks]]
+
+            set remainingBooks $books
+            foreach book $newGroupBooks {
+                set remainingBooks [lremoveFirst $remainingBooks $book]
+            }
+
+            # I hate hardcoding the name of a recursive proc in the proc
+            set procname [lindex [info level 0] 0]
```
Comment (kotp) [inline exercises/book-store/example.tcl:44]:
```suggestion
            set procname [lindex [info level 0] 0]
```

File: exercises/react/react.test [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+        set callback1 [$output addCallback {{cell} {$cell value}}]
+        set callback2 [$output addCallback {{cell} {$cell value}}]
+        $output removeCallback $callback1
+        $output removeCallback $callback1
+        $output removeCallback $callback1
+
+        $input setValue 2
+        list [$callback1 value] [$callback2 value]
+    } -result {{} 3}
+
```
Comment (kotp) [inline exercises/react/react.test:201]:
```suggestion

```

Comment (kotp) [reply exercises/react/react.test:201]: This is to delete trailing space, not the line.

File: exercises/book-store/example.tcl [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+}
+
+interp alias {} basketCost {} BookStore::cost
+
+
+############################################################
+# Some utility procs
+
+# Find the distinct elements of this list, without sorting
+proc ldistinct {list} {
```
Comment (sshine) [inline exercises/book-store/example.tcl:80]: ~Isn't `list` Tcl's special variadic argument sauce, like [Python's `*args`](https://stackoverflow.com/a/36908/235908)?~

No, it's also `args` in Tcl.

File: exercises/react/react.tcl [hunk 5]
```diff
@@ -0,0 +1,17 @@
+# This stub has less detail than some other exercises.
```
Comment (sshine) [inline exercises/react/react.tcl:1]: Is this a note to self? I find that it is somewhat meta and unhelpful in the eyes of the student. Considering the statement, perhaps it should be removed in favor of either nothing, more details in the stub, more details in the hint, or something else?

The more details could be that the student should check the test file to see what methods are missing.

But honestly, I would probably have a hard time figuring out what to add, since filling the gap between an almost empty stub and a test file that uses the OO interface heavily.

I had a comparable bad experience when taking the OCaml course and reaching the "hangman" exercise that uses FRP (functional reactive programming) which I had zero experience with. So I had to learn a paradigm using a library with no simple examples that resembled the exercise, and an exercise that wasn't beginner-friendly.

What I ended up doing was cheating, looking at the example solution and say to myself "Oh, so that's what they want me to do!" and I made it into a "refactoring the example solution" exercise instead, putting thoughts into how to best propagate events.

I suspect that some students might not "cheat" and make the most of it this way.

TL;DR: How about adding `method` stubs? Or place in the hints which `method`s you need to add?

**Comment by glennj** [comment]
> How many side exercises remain in the pipeline for approval?

After these 3 I have 30 ready for PR. https://docs.google.com/spreadsheets/d/19VdB52hf3iRY2U_QpFn54CcLNXp8wlqq98y99X89sUA/edit#gid=1704526598

> We could look at publishing the track soon.
> In starting a new job I've unsubscribed from mentoring on Exercism. Perhaps, if you intend to mentor Tcl, you can let me know if there's any significant traffic. (Mentoring Standard ML is essentially a no-operation.)

I don't expect much traffic. If things start getting hairy, I'll ping you (what's your preferred medium for that?) or recruit some Tclers.

File: exercises/react/react.tcl (hunk 5 above)
Comment (glennj) [reply exercises/react/react.tcl:1]: I'll fix this up, and then merge.

**Comment by sshine** [comment]
Anywhere on GitHub or email is good.
